### PR TITLE
Fix issue #43—Callback url registered incorrectly if using route groups

### PR DESCRIPTION
**Pull request by bgisme**
Attempt to fix issue #43—Callback url registered incorrectly if using route groups.

It adds an optional parameter to `ServiceRegister` called `grouped: [PathComponent]`
```
oAuth<OAuthProvider>(
        grouped: [PathComponent] = [],
        ...
```

Can be used with group routes like this...
```
let abComponents: [PathComponent] = ["a", "b"]
let abRoutes = routes.grouped(abComponents)
try abRoutes.oAuth(
    grouped: abComponents,
    from: Google.self,
    authenticate: "start",
    callback: "http://localhost:8080/a/b/oauth/google",
    scope: ["profile", "email"],
    redirect: "/done")
    
 // call to trigger... http://localhost:8080/a/b/start
```
Ungroued routes can omit the parameter. So existing usage should not change.

The parameter is passed down to FederatedServiceRouter `configureRoutes(grouped:withAuthURL:authenticateCallback:on:)` where the callback URL gets registered.

When breaking the `callbackURL` into path components, duplicate components at the head of the path are removed. So the grouped route is registered properly and matches the callback url.

Most of the work happens inside `ImperialCore` > `Helpers` > `String + Tools.swift` > `pathComponents(grouped:)`

Also added a few tests for different kinds of groups and callback urls.

Happy to add readme stuff if you like the fix.

### CODE REVIEW

File: Tests/ImperialTests/ImperialTests.swift [hunk 1]
```diff
@@ -4,7 +4,20 @@ import XCTest
 class ImperialTests: XCTestCase {
     func testExists() {}
```
Comment (bgisme) [inline Tests/ImperialTests/ImperialTests.swift:6]: Tests for different kinds of grouped scenarios

File: Sources/ImperialCore/Routing/FederatedServiceRouter.swift [hunk 2]
```diff
@@ -77,9 +77,9 @@ extension FederatedServiceRouter {
     public var errorKey: String { "error" }
     public var callbackHeaders: HTTPHeaders { [:] }
    
-    public func configureRoutes(withAuthURL authURL: String, authenticateCallback: ((Request) throws -> (EventLoopFuture<Void>))?, on router: RoutesBuilder) throws {
-		router.get(callbackURL.pathComponents, use: callback)
-		router.get(authURL.pathComponents) { req -> EventLoopFuture<Response> in
+    public func configureRoutes(grouped: [PathComponent], withAuthURL authURL: String, authenticateCallback: ((Request) throws -> (EventLoopFuture<Void>))?, on router: RoutesBuilder) throws {
```
Comment (bgisme) [inline Sources/ImperialCore/Routing/FederatedServiceRouter.swift:80]: Use the new func pathComponents(grouped:) on both callbackURL and authURL

File: Sources/ImperialCore/Helpers/String+Tools.swift [hunk 3]
```diff
@@ -2,16 +2,31 @@ import Foundation
 import RoutingKit
 
 extension String {
```
Comment (bgisme) [inline Sources/ImperialCore/Helpers/String+Tools.swift:4]: Change pathComponents var to a func.
Break url path into components like before.
Remove any duplication starting with first index. Stop at the first index that does not match.

File: Sources/Imperial/Services/DeviantArt/DeviantArt.swift [hunk 4]
```diff
@@ -6,6 +6,7 @@ public class DeviantArt: FederatedService {
 
     @discardableResult
     public required init(
+        grouped: [PathComponent],
```
Comment (bgisme) [inline Sources/Imperial/Services/DeviantArt/DeviantArt.swift:9]: Add parameter `grouped: [PathComponent]`
It gets passed down in to where the `callback` url is parsed into path components and registered.
Any overlap between callback url path components and `grouped` parameter is removed before registration.
So the grouped route matches the callback url.

**Comment by 0xTim** [comment]
Just an FYI I'm going to be doing an Imperial code sprint next week to go through all the issues and PRs

**Comment by 0xTim** [comment]
@fpseverino can we check that this is fixed in 2.0.0 beta?

**Comment by fpseverino** [comment]
Yeas, it looks like it's fixed now. In #109 I switched from using a custom implementation of `pathComponents` to the one provided by RoutingKit, so maybe that was it.
Anyway, I'm opening a PR that adds tests for this

**Pull request closed without merging**